crunch-core/src/main/java/org/apache/crunch/util/CrunchRenameCopyListing.java

```diff
@@ -0,0 +1,269 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information regarding copyright ownership.  The ASF licenses this file to you under the
+ * Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.  You may obtain a
+ * copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS"
+ * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+package org.apache.crunch.util;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.SequenceFile;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.security.Credentials;
+import org.apache.hadoop.tools.CopyListing;
+import org.apache.hadoop.tools.CopyListingFileStatus;
+import org.apache.hadoop.tools.DistCpOptions;
+import org.apache.hadoop.tools.DistCpOptions.FileAttribute;
+import org.apache.hadoop.tools.SimpleCopyListing;
+import org.apache.hadoop.tools.util.DistCpUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Stack;
+
+/**
+ * A custom {@link CopyListing} implementation capable of dynamically renaming
+ * the target paths according to a configured set of values.
+ * <p>
+ * Once https://issues.apache.org/jira/browse/HADOOP-16147 is available, this
+ * class can be significantly simplified.
+ * </p>
+ */
+public class CrunchRenameCopyListing extends SimpleCopyListing {
+  /**
+   * Comma-separated list of original-file:renamed-file path rename pairs.
+   */
+  public static final String DISTCP_PATH_RENAMES = "crunch.distcp.path.renames";
+
+  private static final Logger LOG = LoggerFactory.getLogger(CrunchRenameCopyListing.class);
+  private final Map<String, String> pathRenames;
+
+  private long totalPaths = 0;
+  private long totalBytesToCopy = 0;
+
+  /**
+   * Protected constructor, to initialize configuration.
+   *
+   * @param configuration The input configuration, with which the source/target FileSystems may be accessed.
+   * @param credentials - Credentials object on which the FS delegation tokens are cached. If null
+   * delegation token caching is skipped
+   */
+  protected CrunchRenameCopyListing(Configuration configuration, Credentials credentials) {
+    super(configuration, credentials);
+
+    pathRenames = new HashMap<>();
+
+    String[] pathRenameConf = configuration.getStrings(DISTCP_PATH_RENAMES);
```

Comment:
I was mentioning to @noslowerdna offline, but this strategy of storing the rename mappings in the Configuration could bloat the Configuration a bit.  As we were talking though it, our rough math was that each rename mapping would be something like ~20 bytes (10 bytes original filename plus 10 bytes target filename).  The rename entries don't contain fully qualified paths.  The worst-case jobs I'm used to seeing are on the order of 100,000 map tasks.  In such a scenario, that would be ~2MB of data added to the Configuration.

I don't see that causing any catastrophic problem.  I was going to look to see if there is a limit anywhere as to how big a Configuration is allowed to get.

If this does become a problem, the DISTCP_PATH_RENAMES config key/value could be replaced with a config key/value that points to a file in the Crunch tmp dir that stores the rename mappings.

Reply:
FYI - so far I've been unable to find any mention of a limit on Configuration size.  I did a little googling and looking over mapred-default.xml and yarn-default.xml in case they might have something that would cause rejection of jobs with configs over a size threshold.  For some reason I felt like I might have seen something like that before but perhaps I was thinking of something else.

Reply:
Thanks @ben-roling, when I have a chance next week I'll run a test job with an artificially very large number of renames to see if I can uncover any problems with that approach.

Reply:
I ran some tests and did not encounter any problems I would consider showstopping. My largest concern is it appeared that the DistCp job completion time was delayed roughly proportionally to the size of the configuration, with a delay of nearly two minutes observed in the worst case test - I think probably due to persisting the configuration into job history. 

```
19/03/05 11:04:53 INFO mapreduce.Job:  map 100% reduce 0%
19/03/05 11:06:48 INFO mapreduce.Job: Job job_1538164922410_717171 completed successfully
```

Otherwise everything looked good, and viewing the job configuration through the web UI worked fine although slow when the size was in multiple MBs.

The configuration property value size was retrieved by,
```
conf.get(CrunchRenameCopyListing.DISTCP_PATH_RENAMES).length()
```

Number of part files | Configuration size | Completion delay
-- | -- | --
10k | 260kb | 0:05
50k | 1.3mb | 0:29
250k | 6.8mb | 0:54
500k | 13.8mb | 1:55